### PR TITLE
Refactor error handling + dynamic scheduling in SlackIngestionCommander

### PR DESCRIPTION
**Pull request by leogrim**
@ryanpbrewster @andrewconner

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackIngestionCommander.scala [hunk 1]
```diff
@@ -17,21 +18,26 @@ import com.keepit.common.time._
...
-import scala.util.{ Try, Failure, Success }
+import scala.util.{ Failure, Success }
 
 object SlackIngestionCommander {
-  val delayBetweenIngestions = Period.minutes(10)
-  val retryDelay = Period.minutes(1)
-  val ingestionGracePeriod = Period.minutes(30)
+  val nextIngestionDelayAfterFailure = Period.minutes(5)
+  val nextIngestionDelayWithoutNewMessages = Period.minutes(10)
+  val nextIngestionDelayAfterNewMessages = Period.minutes(1)
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackIngestionCommander.scala:26]: Being more aggressive if we got something the last time we ingested

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackIngestionCommander.scala [hunk 2]
```diff
@@ -17,21 +18,26 @@ import com.keepit.common.time._
...
+
+  val ingestionTimeout = Period.minutes(30)
   val integrationBatchSize = 10
   val messageBatchSize = 25
 
   val slackLinkPattern = """<(.*?)(?:\|(.*?))?>""".r
+
+  case class BrokenSlackIntegration(integration: SlackIntegration, token: Option[SlackAccessToken], cause: Option[SlackAPIFailure]) extends Exception(s"Found a broken Slack integration: token->$token, integration->$integration, cause->$cause")
+  case class ForbiddenSlackIntegration(integration: SlackIntegration) extends Exception(s"Found a forbidden Slack integration: $integration")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackIngestionCommander.scala:35]: You think these should belong in `SlackAPIFailure`?

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackIngestionCommander.scala [hunk 3]
```diff
@@ -50,63 +56,78 @@ class SlackIngestionCommanderImpl @Inject() (
...
-          }
-          Future.successful(())
+      val allIngestedFuture = FutureHelpers.sequentialExec(integrations) {
+        case integration => ingestMaybe(integration, isAllowed, getToken).imap(_ => ()).recover { case _ => () }
+      }
+      allIngestedFuture.imap(_ => integrations.isEmpty)
+    }
+  }
+
+  private def ingestMaybe(integration: SlackChannelToLibrary, isAllowed: Id[SlackChannelToLibrary] => Boolean, getAuthToken: Id[SlackChannelToLibrary] => Option[SlackAccessToken]): Future[Option[SlackMessageTimestamp]] = {
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackIngestionCommander.scala:89]: Is `getAuthToken` supposed to be a `Map`?

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackIngestionCommander.scala (hunk 2 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackIngestionCommander.scala:35]: Well it's a higher above, our data being wrong basically, and I don't want to pollute our `SlackClient`. I could move them to `SlackIntegration` though, if you want to use them the other way around.

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackIngestionCommander.scala (hunk 3 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackIngestionCommander.scala:89]: Doesn't really matter, it's something that provides you with an auth token

**Pull request merged**